File: ios/MegaPDF/DocumentCapabilities.swift
import Foundation

/// What the app lets the user do to the open document, from its security (#131).
/// The mapping is ADR-004 §2, the same on every platform:
///
/// | capability | granted by | what it gates |
/// |---|---|---|
/// | canFillForms | fill forms or annotate | form fields |
/// | canSign | fill forms or annotate | signatures, stamps and check marks |
/// | canAddText | modify, fill forms or annotate | text boxes |
/// | canEditContent | modify | editing the document's own text |
///
/// A form that allows filling lets people fill in everything it offers: its fields,
/// check marks, signatures and text boxes. Changing the document's own text needs
/// modify. Annotate implies form filling (ISO 32000).
///
/// Pure, so the policy is tested without a document.
struct DocumentCapabilities: Equatable {
    /// Retyping or deleting the document's own text.
    let canEditContent: Bool
    /// Placing, moving and removing signatures, and adding or removing check marks.
    let canSign: Bool
    /// Ticking the document's own form fields.
    let canFillForms: Bool
    /// Adding, correcting, restyling, moving and removing text boxes.
    let canAddText: Bool
    /// Setting, changing or removing the password: only an open with full access (ADR-004 §3).
    let canChangeSecurity: Bool
    /// Protected and opened without full access: the app says so and offers the owner password.
    let isRestricted: Bool

    init(security: PdfSecurity) {
        // Full access is the owner's open: it may change the security itself, so it may
        // change anything, whatever the permission bits say.
        let full = security.hasFullAccess
        let modify = full || security.allows(.modify)
        let fillIn = full || security.allows(.fillForms) || security.allows(.annotate)
        canEditContent = modify
        canSign = fillIn
        canFillForms = fillIn
        canAddText = modify || fillIn
        canChangeSecurity = full
        isRestricted = security.isEncrypted && !full
    }

    /// Whether this open may apply `operation` — the backstop behind every gated tool.
    func allows(_ operation: PdfEditOperation) -> Bool {
        switch operation {
        case is BodyTextEditOperation, is BodyTextDeleteOperation,
             is RedactMarkOperation, is MoveRedactionMarkOperation, is ClearRedactionMarksOperation:
            // A redaction removes the document's own content, so it needs the same
            // permission the Redact command itself does (#329).
            return canEditContent
        case is TextBoxOperation, is EditTextBoxOperation, is MoveTextBoxOperation:
            return canAddText
        case is StampOperation, is MoveStampOperation, is MarkOperation:
            return canSign
        case is FieldToggleOperation:
            return canFillForms
        default:
            // An operation this list doesn't know yet needs every permission.
            return canEditContent && canSign && canFillForms && canAddText
        }
    }
}

File: android/app/src/main/java/com/megapdf/android/DocumentCapabilities.kt
package com.megapdf.android

import com.megapdf.engine.PdfPermissions
import com.megapdf.engine.PdfSecurity

/**
 * What the open document lets the user do, from the permissions its security grants this
 * open (#131, ADR-004 decision 2). Pure, so the mapping is tested on the JVM; the view
 * model asks it before every edit, and the toolbar asks it for what to enable.
 */
data class DocumentCapabilities(
    /** The document's own text and text boxes: the modify permission. */
    val canEditContent: Boolean,
    /** Signatures, stamps and check marks: the annotate permission. */
    val canSign: Boolean,
    /** Ticking form fields: fill forms, which annotate also allows. */
    val canFillForms: Boolean,
    /** Setting, changing or removing the password: full access only (decision 3). */
    val canChangeSecurity: Boolean,
    val isEncrypted: Boolean,
    /** Protected, and this open is not the owner's: the restricted notice and Unlock apply. */
    val isRestricted: Boolean,
) {
    /** Whether this open may apply [operation]. An edit this list does not know needs full access. */
    fun allows(operation: PdfEditOperation): Boolean = when (operation) {
        is BodyTextEditOperation, is BodyTextDeleteOperation,
        is TextBoxOperation, is EditTextBoxOperation, is MoveTextBoxOperation -> canEditContent
        is StampOperation, is MoveStampOperation, is MarkOperation -> canSign
        is FieldToggleOperation -> canFillForms
        else -> canChangeSecurity
    }

    companion object {
        fun fromSecurity(security: PdfSecurity): DocumentCapabilities {
            fun may(permission: Int) = security.hasFullAccess || security.allows(permission)
            return DocumentCapabilities(
                canEditContent = may(PdfPermissions.MODIFY),
                canSign = may(PdfPermissions.ANNOTATE),
                canFillForms = may(PdfPermissions.FILL_FORMS) || may(PdfPermissions.ANNOTATE),
                canChangeSecurity = security.hasFullAccess,
                isEncrypted = security.isEncrypted,
                isRestricted = security.isEncrypted && !security.hasFullAccess,
            )
        }

        /** Nothing open, or an unprotected document: everything is allowed. */
        val FULL = fromSecurity(PdfSecurity.UNPROTECTED)
    }
}

/** Which face the Password command's dialog shows (#131, ADR-004 decision 5). */
enum class PasswordCommandMode {
    /** Without full access: say why, and offer the owner password. */
    RESTRICTED,

    /** Unprotected: set one password, which opens the document with every permission. */
    SET,

    /** Protected, with full access: change the password or remove it. */
    CHANGE,
    ;

    companion object {
        fun of(capabilities: DocumentCapabilities): PasswordCommandMode = when {
            !capabilities.canChangeSecurity -> RESTRICTED
            !capabilities.isEncrypted -> SET
            else -> CHANGE
        }
    }
}

/** Why a new password can't be used yet. */
enum class NewPasswordProblem { EMPTY, MISMATCH }

/** Null when [password] may be set: not empty, and typed the same twice. */
fun newPasswordProblem(password: String, confirmation: String): NewPasswordProblem? = when {
    password.isEmpty() -> NewPasswordProblem.EMPTY
    password != confirmation -> NewPasswordProblem.MISMATCH
    else -> null
}
